### ISSUE TITLE
h5 file locked by scans

### ISSUE THREAD

**Issue by dschick** [sardana-org/sardana#1124]
when running scans and saving it to the NXfileRecorder I cannot open the h5 file by another application like PyMCA.
It seems that h5py should support Single Writer and Multiple Readers (SWMR) starting from version 2.5 ( I do have 2.8 installed).
My question is, if Sardana includes the SWMR feature and if not, is it possible?

**Comment by reszelaz** [comment]
Thanks for the report @dschick. We were aware of that and actually I'm surprised there was no issue created yet...

Sardana does not support SWMR yet. @cpascual made already some progress in this aspect but we hit a limitation related to the way how the recorders work in Sardana and how the SWMR feature is supposed to be used, see https://github.com/h5py/h5py/issues/712 for more details.

Now, an idea came to my mind... and what if we define a singleton class wrapping the `h5py.File` object. When the first recorder (the first scan execution) instance use it, it will be created the descriptor in the write mode. The file will stay opened for writing until it gets used by another recorder instance (another scan execution).  Meanwhile it will be possible open and close readers. Whenever we change the file, the singleton instance will also change the `h5py.File`. In principle this should work, right?

**Comment by dschick** [comment]
sorry for pointing at that :)

Good to here that @cpascual is already working on that.

However, I am not sure if I got your suggestion right, since I was talking about reading the file with an external application  like PyMCA or HDF Viewer while Sardana is writing to it.
Your idea seems to be working for writing to the file with multiple recorders?

**Comment by reszelaz** [comment]
> Your idea seems to be working for writing to the file with multiple recorders?

Sorry, I didn't explain it well. In Sardana we create a dedicated instance of a recorder per scan execution. So every scan has its own instance of H5 recorder. Then every time an instance is created, it tries to open a file for writing. If in the meanwhile one opened a reader application, e.g. PyMCA, the subsequent scan can not open for writing anymore...
With the singleton idea we will keep the file opened for writing all the time (until we change the file), even if the H5 recorder instances will be created and destroyed multiple times.

**Comment by dschick** [comment]
okay, so now I understood the concept better. However, this will not help in case I started a 4h scan and then wants to open the h5 file by a reader application like PyMCA for the first time, right?
I need to open the file with the reader after creation but when no scan is running?

Maybe I just save the data in spec and h5 files simultaneously and then I do not bother with the SWMR feature of h5py...

**Comment by reszelaz** [comment]
> I need to open the file with the reader after creation but when no scan is running?

I'm not 100% sure but I think it should be possible to open the file for reading before sardana finished writing to the datasets i.e. before the scan finished. Does it fail in your case? Or it only fails when another scan starts?

> Maybe I just save the data in spec and h5 files simultaneously and then I do not bother with the SWMR feature of h5py...

For the moment this seems as the best workaround, another would be to change the file for every scan..
We were thinking about adding SWMR support anyway. Now with the idea of handling the file with a singleton it will be much easier to implement it.

**Comment by dschick** [comment]
> I'm not 100% sure but I think it should be possible to open the file for reading before sardana finished writing to the datasets i.e. before the scan finished. Does it fail in your case?

Yep, when scans are running I cannot access the h5 file with any reader as they complain that the file is locked.

**Comment by reszelaz** [comment]
> I cannot access the h5 file with any reader as they complain that the file is locked.

Then maybe we need to use the `swmr_mode = True` already to do this? I would need to read the docs to be able to help here. Soon we will get back to this issue.

**Comment by jangarrevoet** [comment]
As far as I remember correctly, for SWMR it is best to create the file with SWMR enabled, close the file. Open it with the writer and readers.

With “readers” like hdfview, they open files read/write as a standard. For hdfview you can change this in the settings.

**Comment by reszelaz** [comment]
Thanks @jangarrevoet for joining this discussion and sharing your experience!

> I cannot access the h5 file with any reader as they complain that the file is locked.

I confirm this. With the current sardana H5 recorder this fails.

I tried the H5 SWMR feature and I think it will work for us as soon as we use the singleton approach. I have made a simple PoC inspired on this [example](http://docs.h5py.org/en/stable/swmr.html).

I use this `swmr_writer.py`:

```python
import time
import h5py
import numpy as np

f = h5py.File("swmr.h5", 'w', libver='latest')

def scan(id_):
    arr = np.array([1,2,3,4])
    dset = f.create_dataset("data{0}".format(id_), chunks=(2,), maxshape=(None,), data=arr)
    f.flush()
    if id_ == 1:
        f.swmr_mode = True
        # Now it is safe for the reader to open the swmr.h5 file
    raw_input("Dataset created, continue?")
    for i in range(10):
        new_shape = ((i+1) * len(arr), )
        dset.resize( new_shape )
        dset[i*len(arr):] = arr
        dset.flush()
        time.sleep(0.5)
        # Notify the reader process that new data has been written
print("First scan")
scan(1)
raw_input("Second scan?")
scan(2)
``` 
And this `swmr_reader.py`:

```python
import h5py
import time

def read_scan(id_):
    f = h5py.File("swmr.h5", 'r', libver='latest', swmr=True)
    dset = f["data{0}".format(id_)]
    while True:
        dset.id.refresh()
        shape = dset.shape
        print( shape )
        time.sleep(0.1)
        if shape[0] == 40:
            break
    f.close()

print("Reading first scan")
read_scan(1)
raw_input("Read second scan?")
read_scan(2)
```

First starting the writer and only after creating the first scan dataset starting the reader.
Whenever the scan finish I continue with the writer, and only after creating the second scan dataset I continue with the reader. This works fine.

I had however problems with refreshing/updating the file in between the scans from the reader. I tried `h5py.File.flush()` but this does not work for me. So I was forced to close the file and open it again to see the second scan dataset. Not sure if this is a known limitation of h5py (or even HDF5). This could be related to https://github.com/h5py/h5py/issues/432. But anyway I think this is not actually a problem for the sardana recorder, but more for the reading applications.

**Comment by dschick** [comment]
@reszelaz thanks for the update.
Can you test the reading with PyMCA instead of your reader script?

**Comment by jkotan** [comment]
Hi @dschick, 
If you want you can use DESY nexus recorder for sardana. Our recorder https://github.com/nexdatas/sardanascanrecorders works in SWMR mode. I have tested it with MCA sardana scan which I view online with lavue https://github.com/jkotan/lavue . The Desy nexus recorder uses as a background python library h5py or python-pni (depending what is installed). When you use the recorder it is good to set 3 additional nexus tango servers to define configuration for your metadata in local DB. In this approach you don't have to type whole configuration  for Snapshot every time you select MG. At Desy by default we have SWMR mode of because SWMR files are not readable by Matlab and IDL as well as you have to use recent version of libhdf5 (the older one installed in debian9 has  a serious bug: the library crashes when you try to create " too many" fields or groups). But you can easily switch it on by setting the  'writer' property of our NXSDataWriter. If you are tomorrow on sardana workshop/tango meeting at DESY we can discuss it this in details

**Comment by cpascual** [comment]
Hi AFAIK, in ALBA we do access h5 files **with read-only viewers** while the file is being recorded. The "only" problem is , as @reszelaz  mentioned, that if the reader is left open, you won't be able to start a new scan because the new writer won't be able to open the file for writing when a reader has it open (and SWMR does not solve this yet).

One related issue is that some old versions of PyMCA attempted to open the nexus files in write rw mode (even if it was only for reading), which was not possible while the scan was being recorded.

A similar case (already mentioned by @jangarrevoet) occurred wiith hdfview, but we worked it around by changing the behaviour globally in its config files

**Comment by dschick** [comment]
hej all,

thanks for all your feedback. It seems that hdf5 has still some limitations for online viewing of the scans.
So either I will try to create a copy of the hdf5 file after every scan or I just work with spec and hdf5 file simultaneously, whereas I will use the hdf5 for storing 2D images.

It seems that there are several individual solutions around which might find their way into the Sardana master branch one day. Shall this issue kept open for this reason? Otherwise I am fine with closing it.

**Comment by cpascual** [comment]
>  Shall this issue kept open for this reason? 

I would say so. It can be closed when some solution is implemented

**Comment by reszelaz** [comment]
Hi all,

To overcome this problem, and to introduce the SWMR mode, we have introduces a concept of the _H5 write session_ as explained in the [user's guide to the `NXscanH5_FileRecorder`](https://sardana-controls.org/users/scan.html#nxscanh5-filerecorder).

Menwhile, silx workaround the issue of file locking by setting the environment variable `HDF5_USE_FILE_LOCKING=FALSE` (see https://github.com/h5py/h5py/issues/712#issuecomment-562980532 and the following two comments). This variable is available from Available from HDF5 1.10.1.

I proceed to close this one.
